### PR TITLE
[FIX] google_calendar: recurrence splitting

### PR DESCRIPTION
**Pull request by yajo**
Imagine a scenario where Odoo and Google calendars are synced. Then, this happens:
1. User goes to Google and changes one event from a recurrence series.
2. When saving, user chooses to modify "this event and all upcoming".
3. Odoo syncs from Google automatically via cron.

<details>

```
2024-07-04 10:02:21,094 25 INFO odoo odoo.addons.google_calendar.models.res_users: Calendar Synchro - Starting synchronization for res.users(15,)                                                             
2024-07-04 10:02:21,568 25 INFO odoo odoo.addons.google_calendar.models.calendar_recurrence_rule: Recurrence #4838 | current rule: FREQ=WEEKLY;WKST=MO;UNTIL=20240620T215959Z;BYDAY=FR | new rule: FREQ=WEEKLY;WKST=MO;UNTIL=20240613T215959Z;BYDAY=FR | remaining: 1 | removed: 1                                                                                                                                          
2024-07-04 10:02:21,583 25 INFO odoo odoo.models.unlink: User #15 deleted calendar.event records with IDs: [1999701]                                                                                          
2024-07-04 10:02:21,586 25 INFO odoo odoo.models.unlink: User #15 deleted mail.followers records with IDs: [3665655, 3665656, 3665657, 3665658, 3665659, 3665660, 3665661, 3665662, 3665768]                  
2024-07-04 10:02:21,919 25 ERROR odoo odoo.addons.google_calendar.models.res_users: [res.users(15,)] Calendar Synchro - Exception : Record does not exist or has been deleted.                                
(Record: calendar.event(1999701,), User: 15) !                                                                                                                                                                
Traceback (most recent call last):                                                                                                                                                                            
  File "/opt/odoo/custom/src/odoo/odoo/﻿[﻿api.py﻿](https://api.py/)﻿", line 997, in get                                                                                                                                              
    cache_value = field_cache[record._ids[0]]                                                                                                                                                                 
KeyError: 1999701                                                                                                                                                                                             
                                                                                                                                                                                                              
During handling of the above exception, another exception occurred:                                                                                                                                           
                                                                                                                                                                                                              
Traceback (most recent call last):                                                                                                                                                                            
  File "/opt/odoo/custom/src/odoo/odoo/﻿[﻿fields.py﻿](https://fields.py/)﻿", line 1161, in __get__                                                                                                                                      
    value = env.cache.get(record, self)                                                                                                                                                                       
  File "/opt/odoo/custom/src/odoo/odoo/﻿[﻿api.py﻿](https://api.py/)﻿", line 1004, in get                                                                                                                                             
    raise CacheMiss(record, field)                                                                                                                                                                            
odoo.exceptions.CacheMiss: 'calendar.event(1999701,).write_date'                                                                                                                                              
                                                                                                                                                                                                              
During handling of the above exception, another exception occurred:                                                                                                                                           
                                                                                                                                                                                                              
Traceback (most recent call last):                                                                                                                                                                            
  File "/opt/odoo/auto/addons/google_calendar/models/﻿[﻿res_users.py﻿](https://res_users.py/)﻿", line 100, in _sync_all_google_calendar                                                                                                    
    user.with_user(user).sudo()._sync_google_calendar(google)                                                                                                                                                 
  File "/opt/odoo/auto/addons/google_calendar/models/﻿[﻿res_users.py﻿](https://res_users.py/)﻿", line 79, in _sync_google_calendar                                                                                                         
    synced_events = self.env['calendar.event'].with_context(write_dates=events_write_dates)._sync_google2odoo(events - recurrences, default_reminders=default_reminders)                                      
  File "/opt/odoo/auto/addons/google_calendar/models/﻿[﻿google_sync.py﻿](https://google_sync.py/)﻿", line 181, in _sync_google2odoo                                                                                                          
    odoo_record_write_date = write_dates.get(﻿[﻿odoo_record.id﻿](https://odoo_record.id/)﻿, odoo_record.write_date)                                                                                                                          
  File "/opt/odoo/custom/src/odoo/odoo/﻿[﻿fields.py﻿](https://fields.py/)﻿", line 1191, in __get__                                                                                                                                      
    raise MissingError("\n".join([                                                                                                                                                                            
odoo.exceptions.MissingError: Record does not exist or has been deleted.                                                                                                                                      
(Record: calendar.event(1999701,), User: 15)
```

</details>

In such case, due to the way this was handled, Odoo would die with `MissingError`.

Here I contribute a test for such scenario and the fix.

@moduon MT-6287





---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/google_calendar/tests/test_sync_google2odoo.py [hunk 1]
```diff
@@ -980,6 +980,195 @@ def test_simple_event_into_recurrency(self):
...
+                    None,
+                    [{"method": "popup", "minutes": 30}],
+                ),
+            ):
+                self.attendee_user.sudo()._sync_google_calendar(self.google_service)
+            events = self.env["calendar.event"].search(
+                [("google_id", "like", google_id)]
+            )
+            self.assertEqual(len(events.exists()), 2)
+
```
Comment (arj-odoo) [inline addons/google_calendar/tests/test_sync_google2odoo.py:1171]: Why are you using those weird patches ?
I think it is easier to do it that way:
value_list = [...]
self.env['calendar.event']._sync_google2odoo(GoogleEvent(value_list))

You know that you will sync from google to odoo. No need to call _sync_google2odoo then _sync_odoo2google in this _sync_google_calendar call. It is more clear , lighter and faster.

File: addons/google_calendar/models/google_sync.py [hunk 2]
```diff
@@ -177,13 +177,17 @@ def _sync_google2odoo(self, google_events: GoogleEvent, default_reminders=()):
...
+                # Use the record's write_date to apply Google updates only if they are newer than Odoo's write_date.
+                odoo_record_write_date = write_dates.get(odoo_record.id, odoo_record.write_date)
+                # Migration from 13.4 does not fill write_date. Therefore, we force the update from Google.
+                if not odoo_record_write_date or updated >= pytz.utc.localize(odoo_record_write_date):
+                    vals = dict(self._odoo_values(gevent, default_reminders), need_sync=False)
+                    odoo_record.with_context(dont_notify=True)._write_from_google(gevent, vals)
+                    synced_records |= odoo_record
+            except MissingError:
+                # The record has been deleted in the meantime. Nothing to sync.
+                pass
```
Comment (arj-odoo) [inline addons/google_calendar/models/google_sync.py:190]: Encapsulating everything in try/except is a bad practice and it has a cost. What I would do is check that the odoo_record exists before using it.
1) `odoo_record = self.browse(gevent.odoo_id(self.env)).exists()`
2) avoid doing things if you know they will fail.
```py
 if not odoo_record:
    continue
```

**Review by arj-odoo — CHANGES_REQUESTED** [review]
Thanks for your PR. I understand your issue and made some suggestions. :-)

**Comment by yajo** [comment]
After fixing this problem in my DB, I found out a new one. Since it's closely related, I took the freedom to include its fix also in this PR. After all, it wouldn't be possible to trigger it unless the previous fix landed.

The new commit is: 

> [FIX] google_calendar: avoid MissingError when cancelling already-removed events
>
> When a user re-schedules a recurring event, it can cause some events to be removed. In that case, they shouldn't be cancelled again in Odoo, as they don't exist anymore.

<details>

<summary>Open to see traceback of this new issue</summary>

```
2024-07-09 10:06:06,838 20 INFO odoo odoo.addons.google_calendar.models.res_users: Calendar Synchro - Starting synchronization for res.users(28,)
2024-07-09 10:06:07,321 20 INFO odoo odoo.models.unlink: User #28 deleted calendar.recurrence records with IDs: [6688]
2024-07-09 10:06:07,346 20 INFO odoo odoo.addons.google_calendar.models.calendar_recurrence_rule: Recurrence 4697 FREQ=WEEKLY;UNTIL=20240624T215959Z;BYDAY=TU has all events (11)  deleted because of base event value change: [('start', '06/04/2024, 08:00:00', '-->', '06/11/2024, 08:00:00'), ('stop', '06/04/2024, 09:00:00', '-->', '06/11/2024, 09:00:00')]
2024-07-09 10:06:07,355 20 INFO odoo odoo.models.unlink: User #28 deleted mail.message records with IDs: [1006276]
2024-07-09 10:06:07,389 20 INFO odoo odoo.models.unlink: User #28 deleted calendar.event records with IDs: [1938622, 1938626, 1938621, 1938619, 1938627, 1938618, 1938620, 1938625, 1938617, 1938623, 1938624]
2024-07-09 10:06:07,394 20 INFO odoo odoo.models.unlink: User #28 deleted mail.followers records with IDs: [3593768, 3593769, 3593770, 3593771, 3593772, 3593773, 3593774, 3593775, 3593776, 3593777, 3593779, 3593780, 3593781, 3593782, 3593783, 3593784, 3593785, 3593786, 3593787, 3593788, 3593789, 3593790, 3593791, 3593792, 3593793, 3593794, 3593795, 3593796, 3593797, 3593798, 3593799, 3593800, 3593801, 3593802, 3593803, 3593804, 3593805, 3593806, 3593807, 3593808, 3593809, 3593810, 3593811, 3593812, 3593813, 3593814, 3593815, 3593816, 3593817, 3593819, 3593820, 3593821, 3593822, 3593823, 3593824, 3593825, 3593826, 3593827, 3593828, 3593829, 3593830, 3593831, 3593832, 3593833]
2024-07-09 10:06:07,529 20 INFO odoo odoo.addons.google_calendar.models.calendar_recurrence_rule: Recurrence #4697 | current rule: FREQ=WEEKLY;COUNT=12;BYDAY=TU | new rule: FREQ=WEEKLY;UNTIL=20240624T215959Z;BYDAY=TU | remaining: 2 | removed: 0
2024-07-09 10:06:07,558 20 ERROR odoo odoo.addons.google_calendar.models.res_users: [res.users(28,)] Calendar Synchro - Exception : Record does not exist or has been deleted.
(Record: calendar.event(1938625,), User: 28) !
Traceback (most recent call last):
  File "/opt/odoo/custom/src/odoo/odoo/﻿﻿api.py﻿﻿", line 997, in get
    cache_value = field_cache[record._ids[0]]
KeyError: 1938625

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/odoo/custom/src/odoo/odoo/﻿﻿fields.py﻿﻿", line 1161, in __get__
    value = env.cache.get(record, self)
  File "/opt/odoo/custom/src/odoo/odoo/﻿﻿api.py﻿﻿", line 1004, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: 'calendar.event(1938625,).user_id'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/odoo/auto/addons/google_calendar/models/﻿﻿res_users.py﻿﻿", line 100, in _sync_all_google_calendar
    user.with_user(user).sudo()._sync_google_calendar(google)
  File "/opt/odoo/auto/addons/google_calendar/models/﻿﻿res_users.py﻿﻿", line 79, in _sync_google_calendar
    synced_events = self.env['calendar.event'].with_context(write_dates=events_write_dates)._sync_google2odoo(events - recurrences, default_reminders=default_reminders)
  File "/opt/odoo/auto/addons/google_calendar/models/﻿﻿google_sync.py﻿﻿", line 173, in _sync_google2odoo
    cancelled_odoo._cancel()
  File "/opt/odoo/auto/addons/google_calendar/models/﻿﻿calendar.py﻿﻿", line 325, in _cancel
    my_cancelled_records = self.filtered(lambda e: e.user_id == user)
  File "/opt/odoo/custom/src/odoo/odoo/﻿﻿models.py﻿﻿", line 5444, in filtered
    return self.browse([﻿﻿rec.id﻿﻿ for rec in self if func(rec)])
  File "/opt/odoo/custom/src/odoo/odoo/﻿﻿models.py﻿﻿", line 5444, in <listcomp>
    return self.browse([﻿﻿rec.id﻿﻿ for rec in self if func(rec)])
  File "/opt/odoo/auto/addons/google_calendar/models/﻿﻿calendar.py﻿﻿", line 325, in <lambda>
    my_cancelled_records = self.filtered(lambda e: e.user_id == user)
  File "/opt/odoo/custom/src/odoo/odoo/﻿﻿fields.py﻿﻿", line 2804, in __get__
    return super().__get__(records, owner)
  File "/opt/odoo/custom/src/odoo/odoo/﻿﻿fields.py﻿﻿", line 1191, in __get__
    raise MissingError("\n".join([
odoo.exceptions.MissingError: Record does not exist or has been deleted.
(Record: calendar.event(1938625,), User: 28)
```

</details>

**Comment by yajo** [comment]
All attended! I modified a bit your suggestion, but the essence is the same.

**Review by arj-odoo — APPROVED** [review]

**Comment by arj-odoo** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**